docs/architecture/adr/ADR040-cluster-stability-node-replacement.md

```diff
@@ -0,0 +1,68 @@
+# ADR040: Ensuring cluster stability while replacing nodes
+
+## Status
+
+Pending
+
+## Context
+
+SREs sometimes need to change various things about a cluster's worker nodes, and therefore replace all of the nodes:
+* AMI (e.g. for an EKS update)
+* instance type
+* anything else in the launch template in some way (e.g. instance role)
+In addition to this, in our Terraform we get the latest AMI ID for our EKS version from AWS - this means that as soon as AWS releases a new AMI for our current EKS version, the next time the cluster's deployer step runs for Terraform it will also replace all the nodes.
+
+We have a CloudFormation UpdatePolicy on our worker node auto-scaling-groups that tells it to replace 1 node at a time in each ASG. This would be fine if we didn't have 3 ASGs with a small number of nodes in each - right now this practice terminates too many instances at once and results in the cluster being very unhappy.
+
+Without such a policy we'd be able to update the launch template for the ASG and it wouldn't remove any existing nodes, just set up any new future nodes correctly.
+
+The reason we have 3 small ASGs is so the cluster auto-scaler can scale nodes independently in each AZ - it will know that a pod attempting (and failing) to schedule is tied to a persistent volume claim in a particular availability zone and can scale up the ASG for that zone.
+
+## Possible actions
+
+Here are some of the actions we may want to consider. It is expected we'll choose some subset/variation of these.
+
+### Action 1
+
+Remove this CloudFormation UpdatePolicy and find some alternative to the update policy that ensures the cluster eventually fully rolls out the expected launch template.
+
+#### A
+
+Manual cordoning, draining and termination of nodes
+
+Cons:
+* this would make it some rather irritating toil as an engineer would have to manually run the process
+
+#### B
+
+A script that we run that cordons, drains and terminates nodes - triggered manually.
+
+Cons:
+* we'd have to notice the new launch template and trigger it manually
+
+#### C
+
+A script that runs regularly that terminates nodes - this would ensure the cluster is eventually updated, just not immediately.
+
+Cons:
+* If this was a CronJob inside the cluster, it might terminate itself while running and leave us with an inconsistent cluster? But if it's a Lambda maybe not. Or if it's written to only kill one node and wait until the next run to kill another we might also be okay.
+
+### Action 2
+
+A lifecycle hook in the ASG that tells it to drain nodes etc. when it terminates them. This might have some benefit if we want to e.g. use spot instances in the future?
+
+Cons:
+* nodes might still get terminated too quickly for the cluster to cope
+
+### Action 3
+
+Scale autoscaler down so it's not running while we replace nodes
```

Comment:
We should add that this is AWS' recommended approach in their documentation (and maybe link to it: https://docs.aws.amazon.com/eks/latest/userguide/update-stack.html)